### ISSUE TITLE
Calling print() on array containing itself crashes without error messages

### ISSUE THREAD

**Issue by ghost** [godotengine/godot#8693]
**Operating system or device - Godot version:**
Linux x86_64 / Godot 3.0 alpha

**Issue description:**

Calling `hash()` on an array containing itself crashes the game without error messages, possibly caused by an infinite loop in the hash function.
Expected: printed hash value, debugger breaking the program, or continue with error message
Happened: crash without error messages

**Steps to reproduce:**
1. Create new project & scene
2. Create new Node and attach this GDScript:

```
extends Node

func _ready():
	var arr = []
	arr.append(arr)
	print(arr.hash())
```

3. Save & run the scene

**Minimal project:**
[hasharray.zip](https://github.com/godotengine/godot/files/982617/hasharray.zip)

**Comment by Zylann** [comment]
Printing an array containing itself also crashes.

**Comment by sergey-push** [comment]
we can use "[Floyd's cycle finding algorithm](https://en.wikipedia.org/wiki/Cycle_detection#Tortoise_and_hare)" to detect circular references. It has a linear complexity with no additional space used thus shouldn't impact performance.
The idea is that you "launch" 2 pointers with different speed (1 and 2), first one you use for operations, second one for cycle detection. If there is a cycle somewhere the 2 will meet.

**Comment by reduz** [comment]
this is waaaaaaaaaaaaaaaaaaaaaaaaaay too difficult to avoid, maybe those functions should have a maximum recursion number. For now kicking to 3.1 since itś not critical..

**Comment by Zylann** [comment]
But... why not simply consider the array's reference as being the hash? Do we reaaally want the content to be used for the hash, even though it could not matter at all in many cases? (see my comment above)

**Comment by sergey-push** [comment]
Or we could do as nodejs does with console.log, print only 1-2 levels.

**Comment by Zylann** [comment]
Ah, for console output it could maybe be a different issue actually, since the chosen solution could differ and it's a different use case.

**Comment by reduz** [comment]
It used to be this way, but was changed to this because it made it
impossible to use arrays as dictionary keys, which is something that was
requested.

As I said, fix is a depth counter in functions that do this kind of stuff
so we stop at, say, 1000 and print an error.. but given this is not really
critical will be pushed back

**Comment by Zylann** [comment]
I'm curious why it prevents using as dictionary keys, since other languages work this way by default (and providing ways of customizing it if needed)

**Comment by reduz** [comment]
Because you put [1,2,3] as key, then create another identical array to
index. The hashes will differ and wont be found

**Comment by duurkak** [comment]
on macOs High Sierra 10.13.3 (17D102)
Godot suddenly crashing with "reopen" dialog. Because of these;
```
arr.append(arr)
print(arr)
```
```
arr.push_back(arr)
print(arr)
```

```
arr.push_front(arr)
print(arr)
```

![crash](https://user-images.githubusercontent.com/36481442/36571176-7c3c5626-1846-11e8-9b9f-5aaab8c6383d.png)


it's fine `arr += arr`

**Comment by Chaosus** [comment]
Closed as duplicate of #1289
